### PR TITLE
Flaggable effects

### PR DESCRIPTION
**Pull request by alkemann**
#### Summary
```SUMMARY: Mods "Effect types can have flags"```

#### Purpose of change
To support more JSON and modding contributions, effects should be less hardcoded. We can check creatures by `effect_type_id`, but since Effect Type includes flavor text (name and description shown in interface) a more useful functionality is that several effect types can have the same effect (meaning a single check in code). Flags seems a consistent way of doing this.

#### Describe the solution
The `effect_type` JSON object is given the ability to include flags. Creature class is given a `has_effect_with_flag` method that can be used similar places as `has_effect` or the `Character:has_worn_with_flag`.

#### Describe alternatives you've considered
Going deeper down the rabbit hole and replacing the use of `has_effect( effect_id("id") )`, but that will be looked at in a later PR.

#### Additional context
Will be used by PR #33662 . The actual effect_type and and spell in the changeset is for testing purposes and will be removed before merge (but added back in in #33662 in the Magiclysm mod.

### CODE REVIEW

File: src/creature.cpp [hunk 1]
```diff
@@ -1101,6 +1101,22 @@ bool Creature::has_effect( const efftype_id &eff_id, body_part bp ) const
...
 }
 
+bool Creature::has_effect_with_flag( const std::string &flag, body_part bp ) const
+{
+    if( bp != num_bp ) {
+        return false; // only support general effects for now
+    }
+    for( auto &elem : *effects ) {
+        for( auto &_it : elem.second ) {
+            effect e = _it.second;
```
Comment (BevapDin) [inline src/creature.cpp]:
```suggestion
            const effect &e = _it.second;
```

File: src/effect.h [hunk 2]
```diff
@@ -93,6 +93,8 @@ class effect_type
         int int_decay_tick;
         time_duration int_dur_factor;
 
+        std::vector<std::string> flags;
```
Comment (BevapDin) [inline src/effect.h]:
```suggestion
        std::set<std::string> flags;
```
Doesn't make sense to have multiple entries with the same value and it checking for presence of a specific value is faster. Consequently, you can load this from JSON via `flags = jo.get_tags( "flags" )`.

File: src/effect.cpp [hunk 3]
```diff
@@ -1294,8 +1294,19 @@ void load_effect_type( JsonObject &jo )
 
     new_etype.impairs_movement = hardcoded_movement_impairing.count( new_etype.id ) > 0;
 
+    if( jo.has_member( "flags" ) ) {
```
Comment (BevapDin) [inline src/effect.cpp]:
```suggestion
    flags = jo.get_tags( "flags" );
```

File: src/creature.h [hunk 4]
```diff
@@ -345,6 +345,8 @@ class Creature
         /** Check if creature has the matching effect. bp = num_bp means to check if the Creature has any effect
          *  of the matching type, targeted or untargeted. */
         bool has_effect( const efftype_id &eff_id, body_part bp = num_bp ) const;
+        /** Check if createure has the effect by effect flag */
```
Comment (BevapDin) [inline src/creature.h]:
```suggestion
        /** Check if creature has any effect with the given flag. */
```

File: src/creature.cpp [hunk 5]
```diff
@@ -1101,6 +1101,19 @@ bool Creature::has_effect( const efftype_id &eff_id, body_part bp ) const
     }
 }
 
+bool Creature::has_effect_with_flag( const std::string &flag, body_part bp ) const
+{
+    // auto: const std::pair<efftype_id, std::unordered_map<body_part, effect>>
```
Comment (BevapDin) [inline src/creature.cpp]:
```suggestion
```

File: data/json/flags.json [hunk 6]
```diff
@@ -64,6 +64,15 @@
...
     "info": "This gear <bad>prevents</bad> you from <info>seeing</info> anything."
   },
+  {
+    "id": "EFFECT_NIGHT_VISION",
+    "type": "json_flag",
+    "context": [ "EFFECT" ],
+    "conflicts": [ "BLIND" ],
+    "description": "Allows normal sight without external light source.",
+    "//": "Supported by color in propeties",
+    "properties": { "color": "green" }
```
Comment (BevapDin) [inline data/json/flags.json]: What is that for? "properties" are only loaded for item types. On effects, it's ignored.

File: data/json/flags.json [hunk 7]
```diff
@@ -64,6 +64,15 @@
     "description": "Blinds the wearer while worn, and provides nominal protection vs flashbang flashes.",
     "info": "This gear <bad>prevents</bad> you from <info>seeing</info> anything."
   },
+  {
+    "id": "EFFECT_NIGHT_VISION",
+    "type": "json_flag",
```
Comment (BevapDin) [inline data/json/flags.json]: The data loaded from type "json_flag" is not used by the effect system. Have you tried that different values within this object will yield different output from the game?

Comment (alkemann) [reply data/json/flags.json]: This PR adds this new flag type, moving toward the effect system using it more instead of the effect id is the point of the PR?

Comment (BevapDin) [reply data/json/flags.json]: I get the purpose of the PR, however I doubt that this JSON data here (the entry in this file) has any impact on the game.

Where in the game will a player see the content of this entry? Does this somehow change the display in the effect list in the `@` character information screen? I assumed that "json_flag" are only used by the item information dialog.

Comment (alkemann) [reply data/json/flags.json]: This flag can be used by anything that can cause effect to provide night vision in a way that also displays on the @ screen yes.

Comment (BevapDin) [reply data/json/flags.json]: `json_flag::load` does not load  "properties" nor "context". So how can those values be used within the game?

Your new function `Creature::has_effect_with_flag` is only called from `has_nv`, but not from the code that display character information (include the code that displays current effects), and the ` effect_type::flags` set is protected any not exported to that code. That part of the program does not know that effects now have flags. It therefor wont check any `json_flag` definition.

I tested this branch and changed the "description" of this JSON entry to "FOOBAR", but I could not find any "FOOBAR" string within the game. So where is it actually displayed? Also changing `"color": "green"` to `"color": "foobar"` (certainly an invalid color) does not yield any debug message or error.

File: src/effect.cpp (hunk 3 above)
Comment (alkemann) [reply src/effect.cpp]: As far as I can see, JsonObject does not have a `get_tags` method?

File: data/json/flags.json (hunk 7 above)
Comment (alkemann) [reply data/json/flags.json]: yes, the properties as an attempt at future proofing that isnt part of this PR, so I will remove that. The context and description I just followed the conention of the other examples. You are saying that json_flag does not use context? Even though all the other ones here has that?

File: src/effect.cpp (hunk 3 above)
Comment (BevapDin) [reply src/effect.cpp]: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/json.h#L708

File: data/json/flags.json (hunk 7 above)
Comment (BevapDin) [reply data/json/flags.json]: > You are saying that json_flag does not use context? 

Nope, this is where it's loaded: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/flag.cpp#L18 and there  is no mention of "context" there.

> Even though all the other ones here has that?

Yes. The file has it since it was first created (around 841109ad6544fd3659a41a21021f31f9f0501931), but it was never loaded into the program.

Comment (alkemann) [reply data/json/flags.json]: So should I just remove all contexts then?

Comment (BevapDin) [reply data/json/flags.json]: > So should I just remove all contexts then?

If you want to, do so in another PR as it has nothing to do with the content of this PR.

Comment (alkemann) [reply data/json/flags.json]: I appreciate your CRs, many have been helpful and educational, but I don't really understand what you want from me here. To answer the first question you put, only the FLAG name and the fact that it is a 'json_flag' matters. The flag wont appear in game, because the whole point is that the in game parts of the effect, the flavor, goes on new effects that uses this flag. This way many different flavors can have the same effect.

Comment (BevapDin) [reply data/json/flags.json]: I'm not opposed to the PR itself. Adding flags to effect types is fine.

 I'm just opposed to this single JSON entry because *it does not do anything*. Entries with "json_flag" type apply to *item* flags, but not to effect type flags, so this will show something in the item information window, but it won't affect the display of effects. What does it have to do with this PR (which concerns flags of effect types)?

This entry affects the display of *items* that have the "EFFECT_NIGHT_VISION" flag, but you did not add any such item nor are there any such items in the game right now.

----

In other words: this JSON entry does not affect the game at all. It is about as useful as this:
```JSON
{
    "id": "FOOBAR_DDDD",
    "type": "json_flag",
    "context": [ "SOMETHING....." ]
}
```

It's essentially dead data. *Why* is it there?

Comment (alkemann) [reply data/json/flags.json]: I added the context to follow the convention of existing JSON FLAGS. This flag is not suppose to be used with items, just Effects, hence the context of EFFECT and absence of item flags. 

My idea is that a single flag could be combined for items and spells that give the same effect, but I don't know enough at this point on how the item system work. There would have to be a even of "puts item on" that you could catch and add the effect with this flag on the player. This seems beyond the scope of this PR though.

**Comment by alkemann** [comment]
@BevapDin i removed the context from the flag i added, and since #33662 got merged in first, I have adjusted this PR by removing the demo effect and spell and now the flag is the only thing in main, and the usage is exclusively in Magiclysm.
